modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/stat/IgniteStatisticsInMemoryStoreImpl.java

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2020 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.ignite.internal.processors.query.stat;
+
+import org.apache.ignite.IgniteLogger;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+/**
+ * Sql statistics storage in metastore.
+ * Will store all statistics related objects with prefix "stats."
+ * Store only partition level statistics.
+ */
+public class IgniteStatisticsInMemoryStoreImpl implements IgniteStatisticsStore {
+    /** Table -> Partition -> Partition Statistics map, populated only on server nodes without persistence enabled. */
+    private final Map<StatsKey, Map<Integer, ObjectPartitionStatisticsImpl>> partsStats = new ConcurrentHashMap<>();
+
+    /** Logger. */
+    private final IgniteLogger log;
+
+    /**
+     * Constructor.
+     *
+     * @param logSupplier Logger getting function.
+     */
+    public IgniteStatisticsInMemoryStoreImpl(Function<Class<?>, IgniteLogger> logSupplier) {
+        this.log = logSupplier.apply(IgniteStatisticsInMemoryStoreImpl.class);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void clearAllStatistics() {
+        partsStats.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void replaceLocalPartitionsStatistics(
+            StatsKey key,
+            Collection<ObjectPartitionStatisticsImpl> statistics
+    ) {
+        partsStats.put(key,buildStatisticsMap(key, statistics));
+    }
+
+    /** {@inheritDoc} */
+    @Override public Collection<ObjectPartitionStatisticsImpl> getLocalPartitionsStatistics(StatsKey key) {
+        Map<Integer, ObjectPartitionStatisticsImpl> objStatisticsMap = partsStats.get(key);
+
+        return (objStatisticsMap == null) ? Collections.emptyList() : objStatisticsMap.values();
```

Comment:
Returning a projection of the original map could lead to an error since it is possible to clear the collection outside the store